internal/functions/publishkeys/publish-keys.go

```diff
@@ -0,0 +1,133 @@
+package publishkeys
+
+import (
+	"bytes"
+	"encoding/json"
+	"github.com/covid19cz/erouska-backend/internal/constants"
+	"github.com/covid19cz/erouska-backend/internal/logging"
+	"github.com/covid19cz/erouska-backend/pkg/api/v1"
+	"go.uber.org/zap"
+	"net/http"
+	"os"
+)
+
+//PublishKeys Handler
+func PublishKeys(w http.ResponseWriter, r *http.Request) {
+	var ctx = r.Context()
+	logger := logging.FromContext(ctx)
+
+	var request v1.PublishKeysRequestDevice
+
+	dec := json.NewDecoder(r.Body)
+	err := dec.Decode(&request)
+	if err != nil {
+		logger.Debugf("Could not deserialize request from device: %v", err)
+		http.Error(w, "Could not deserialize", http.StatusBadRequest)
+		return
+	}
+
+	logger.Debugf("Handling PublishKeys request: %+v", request)
+
+	var serverRequest = toServerRequest(&request)
+
+	serverResponse, err := passToKeyServer(logger, serverRequest)
+	if err != nil {
+		logger.Debugf("Could not obtain response from Key server: %v", err)
+		return
+	}
+
+	logger.Debugf("Received response from Key server: %+v", serverResponse)
+
+	if serverResponse.Code == "" && serverResponse.ErrorMessage == "" {
+		logger.Infof("Successfully uploaded %v keys to Key server (%v keys sent)", serverResponse.InsertedExposures, len(serverRequest.Keys))
+
+		err = handleKeysUpload(logger, request)
+		if err != nil {
+			logger.Warnf("Error while processing keys upload: %v", err)
+		}
+	} else {
+		// error has occurred!
+		logger.Infof("Key server has refused the keys; code %v, message '%v'", serverResponse.Code, serverResponse.ErrorMessage)
+	}
+
+	sendResponseToClient(logger, w, toDeviceResponse(serverResponse))
+}
+
+func handleKeysUpload(logger *zap.SugaredLogger, request v1.PublishKeysRequestDevice) error {
+	// TODO save to DB
+
+	return nil
+}
+
+func passToKeyServer(logger *zap.SugaredLogger, request *v1.PublishKeysRequestServer) (*v1.PublishKeysResponseServer, error) {
+	blob, err := json.Marshal(request)
+	if err != nil {
+		logger.Debugf("Could not serialize request for Key server: %v", err)
+		return nil, err
+	}
+
+	// TODO
+	keyServerURL := constants.KeyServerURL
+	url, exists := os.LookupEnv("KEY_SERVER_URL")
+	if exists {
+		keyServerURL = url
+	}
```

Comment:
this would be better off passed as a KeyServerConfig object with URL in its struct, the URL passed in as env variable, something like here https://github.com/google/exposure-notifications-verification-server/blob/b54fb0cfc41dfce042ee1b34d1a5754d70b35342/pkg/database/config.go#L31

Reply:
otherwise it is going to be quite hard to unit test:)